### PR TITLE
[SYCL][L0][Plugin] Add support for batching copy commands

### PR DESCRIPTION
**Pull request by asudarsa**
Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

### CODE REVIEW

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 1]
```diff
@@ -903,11 +903,13 @@ typedef struct CommandListBatchConfig {
...
 } zeCommandListBatchConfig;
 
-static const zeCommandListBatchConfig ZeCommandListBatch = [] {
+static const zeCommandListBatchConfig ZeCommandListBatchConfig(bool IsCopy) {
   zeCommandListBatchConfig Config{}; // default initialize
 
   // Default value of 0. This specifies to use dynamic batch size adjustment.
-  const auto BatchSizeStr = std::getenv("SYCL_PI_LEVEL_ZERO_BATCH_SIZE");
+  const auto BatchSizeStr =
+      (IsCopy) ? std::getenv("SYCL_PI_LEVEL_ZERO_COPY_BATCH_SIZE")
```
Comment (keryell) [inline sycl/plugins/level_zero/pi_level_zero.cpp:911]:
```suggestion
      IsCopy ? std::getenv("SYCL_PI_LEVEL_ZERO_COPY_BATCH_SIZE")
```
Curious that all your `c ? a : b` have parenthesis around `c` everywhere.

Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: Thanks for the correction. I changed it everywhere.

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 2]
```diff
@@ -903,11 +903,13 @@ typedef struct CommandListBatchConfig {
   bool dynamic() const { return Size == 0; }
 } zeCommandListBatchConfig;
 
-static const zeCommandListBatchConfig ZeCommandListBatch = [] {
+static const zeCommandListBatchConfig ZeCommandListBatchConfig(bool IsCopy) {
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:906]: This was a static variable with the goal to be initialized only once, and you made it into a function that will be executed every time you query config. Please turn this into 2 static variables initialized once (re-use initialization code).

File: sycl/plugins/level_zero/pi_level_zero.hpp [hunk 3]
```diff
@@ -672,51 +674,59 @@ struct _pi_queue : _pi_object {
...
+    // Approximate number of commands that are allowed to be batched for
+    // this queue.
+    // Added this member to the queue rather than using a global variable
+    // so that future implementation could use heuristics to change this on
+    // a queue specific basis. And by putting it in the queue itself, this
+    // is thread safe because of the locking of the queue that occurs.
+    pi_uint32 QueueBatchSize = {0};
+  } command_batch;
+
+  command_batch ComputeCommandBatch, CopyCommandBatch;
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.hpp:705]: Please add comments

File: sycl/plugins/level_zero/pi_level_zero.hpp [hunk 4]
```diff
@@ -672,51 +674,59 @@ struct _pi_queue : _pi_object {
...
+
+  // Returns true if any commands for this queue are allowed to
+  // be batched together.
+  // For copy commands, IsCopy is set to 'true'.
+  // For non-copy commands, IsCopy is set to 'false'.
+  bool isBatchingAllowed(bool IsCopy);
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.hpp:711]:
```suggestion
  bool isBatchingAllowed(bool IsCopy) const;
```

File: sycl/plugins/level_zero/pi_level_zero.hpp [hunk 5]
```diff
@@ -738,9 +756,20 @@ struct _pi_queue : _pi_object {
...
-  // and ZeOpenCommandListSize.
-  pi_result executeOpenCommandList();
+  // close it, execute it, and reset the corresponding OpenCommandList.
+  // If IsCopy is 'true', then the OpenCommandList containing copy commands is
+  // executed. Otherwise OpenCommandList containing compute commands is
+  // executed.
+  pi_result executeOpenCommandList(bool IsCopy);
+
+  // Wrapper function to execute both OpenCommandLists (Copy and Compute).
+  pi_result executeOpenCommandLists() {
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.hpp:766]: To make it more distinct from the `executeOpenCommandList`:
```suggestion
  pi_result executeAllOpenCommandLists() {
```

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.hpp:766]: Also please add comment about how one should choose between the 2 versions. I.e. in what circumstances both batches need to be closed.

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 6]
```diff
@@ -2817,6 +2831,15 @@ pi_result piContextRelease(pi_context Context) {
   return ContextReleaseHelper(Context);
 }
 
+// This function will initialize batch sizes for the compute anc copy command
+// lists
+static void initializeQueueBatchSizes(pi_queue Queue) {
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:2836]: please move _pi_queue ctor implementation to pi_level_zero.cpp, and remove this function

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 7]
```diff
@@ -4838,9 +4868,16 @@ pi_result piEventGetInfo(pi_event Event, pi_event_info ParamName,
       // Only do the execute of the open command list if the event that
       // is being queried and event that is to be signalled by something
       // currently in that open command list.
-      if (Event->Queue->hasOpenCommandList() &&
-          Event->Queue->OpenCommandList->first == Event->ZeCommandList) {
-        if (auto Res = Event->Queue->executeOpenCommandList())
+      if (Event->Queue->hasOpenCommandList(false /* IsCopy */) &&
+          Event->Queue->ComputeCommandBatch.OpenCommandList->first ==
+              Event->ZeCommandList) {
+        if (auto Res = Event->Queue->executeOpenCommandList(false /* IsCopy */))
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:4874]: maybe 

```using IsCopy = bool;```

and then you can write arguments like

```IsCopy{false}```

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 8]
```diff
@@ -5523,8 +5555,15 @@ static pi_result enqueueMemCopyHelper(pi_command_type CommandType,
 
   // Get a new command list to be used on this call
   pi_command_list_ptr_t CommandList{};
-  if (auto Res = Queue->Context->getAvailableCommandList(Queue, CommandList,
-                                                         PreferCopyEngine))
+  bool UseCopyEngine =
+      (!(Queue->isInOrderQueue()) || UseCopyEngineForInOrderQueue) &&
+      PreferCopyEngine && Queue->Device->hasCopyEngine();
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:5560]: Let's make it into utility ```pi_queue::useCopyEngine(bool PreferCopyEngine = true)```

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 9]
```diff
@@ -5551,7 +5590,8 @@ static pi_result enqueueMemCopyHelper(pi_command_type CommandType,
           pi_cast<std::uintptr_t>(ZeEvent));
   printZeEventList(WaitList);
 
-  if (auto Res = Queue->executeCommandList(CommandList, BlockingWrite))
+  if (auto Res =
+          Queue->executeCommandList(CommandList, BlockingWrite, OkToBatch))
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:5594]: What are the commands that we still don't batch? Maybe we should just make the last argument (OkToBatch) true by default now?

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 10]
```diff
@@ -5812,7 +5862,7 @@ enqueueMemFillHelper(pi_command_type CommandType, pi_queue Queue, void *Ptr,
 
   // Execute command list asynchronously, as the event will be used
   // to track down its completion.
-  if (auto Res = Queue->executeCommandList(CommandList))
+  if (auto Res = Queue->executeCommandList(CommandList, false, OkToBatch))
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:5865]:
```suggestion
  if (auto Res = Queue->executeCommandList(CommandList, UseCopyEngine, OkToBatch))
```

**Review by smaslov-intel — CHANGES_REQUESTED** [review]
LGTM overall, few minor suggestions

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 9 above)
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.cpp:5594]: We do have considerable number of call-sites where this argument is 'false' (wait commands, membuffermap, memadvise, etc). So, we don't need to change the default now.

File: sycl/plugins/level_zero/pi_level_zero.hpp (hunk 3 above)
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.hpp:706]: Done. Thanks

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 2 above)
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: Done. Thanks. I agree. Multiple calls is more expensive here.

File: sycl/plugins/level_zero/pi_level_zero.hpp (hunk 4 above)
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.hpp]: Done as part of bigger commit.

File: sycl/plugins/level_zero/pi_level_zero.hpp (hunk 5 above)
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.hpp]: Done. Thanks.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 6 above)
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: Done. Thanks.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 7 above)
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: Done. Thanks

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 8 above)
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: Done. Thanks

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 11]
```diff
@@ -973,6 +972,94 @@ static const zeCommandListBatchConfig ZeCommandListBatchConfig(bool IsCopy) {
     }
   }
   return Config;
+}();
+
+// Static variable that holds batch config info for copy command batching.
+static const zeCommandListBatchConfig ZeCommandListBatchCopyConfig = [] {
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:978]: please re-use the initialization lambda from the compute config

**Review by smaslov-intel — COMMENTED** [review]
LGTM, just one comment to re-use code for env var inititalization

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 11 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:978]: I mean turn lambda into a named function and us it to init both of configs

Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.cpp:999]: Got it....I made the appropriate changes. Thanks

**Review by smaslov-intel — APPROVED** [review]

**Review by bader — APPROVED** [review]
sycl/doc/EnvironmentVariables.md looks good to me.

**Pull request merged**